### ISSUE TITLE
At login, @knaw.nl resolves to a different KNAW institute

### ISSUE THREAD

**Issue by FlorisFokkinga** [SURFscz/SBS#1491]
So the name, here shown as 'eduid.nl', is wrong.
<img width="992" alt="image" src="https://github.com/SURFscz/SBS/assets/80269262/453b9e8a-92b8-4869-bb20-3b13d24695ac">

**Comment by oharsta** [comment]
@FlorisFokkinga When I look up this user in the SBS test database, I see the following schac_home:
```
mysql> select uid, name, email, schac_home_organisation from users where name like '%SRAM%';
+----------------------------------------------------------------+-----------+----------------------+-------------------------+
| uid                                                            | name      | email                | schac_home_organisation |
+----------------------------------------------------------------+-----------+----------------------+-------------------------+
| c46818640d83ec0355837ba66624b24a90ab619d@acc.sram.eduteams.org | SRAM Demo | sramdemo@hotmail.com | eduid.nl                |
+----------------------------------------------------------------+-----------+----------------------+-------------------------+
```
The schac_home organisation `eduid.nl` does not resolve to a description in the https://metadata.surfconext.nl/signed/2023/edugain-downstream-idp.xml feed, and therefore the schac_home is show. What were you expecting and why?

**Comment by FlorisFokkinga** [comment]
Sorry, the screenshot is only an example. The problem is with domain knaw.nl.

**Comment by oharsta** [comment]
This is caused by the use of a regular expression for the Scope. Currently in de edugain feed of ~25MB there is one use of a regexp in the Scope element:
```
curl 'httpscurl 'https://metadata.surfconext.nl/signed/2023/edugain-downstream-idp.xml' | grep 'regexp="1"'
<shibmd:Scope regexp="1">.*\.knaw\.nl$</shibmd:Scope>
```
Currently we don't support regular expressions in the `shibmd:Scope` element. In the future we might.

**Comment by oharsta** [comment]
Fix made in https://github.com/SURFscz/SBS/commit/329a704712df4bcf0cb19077001cd904755ced8d#diff-746877e1e4e470b1adee1436c6c7d330a23ae29b2f65a6234479bf6792363a13. Pending on merge branch.

**Comment by oharsta** [comment]
We now parse all the reg-exp `shibmd:Scope` and as a fall-back check each domain against the reg-exp's to resolve IdP display names. See the meta-data parser CRON job in the system tab. 

![Image](https://github.com/SURFscz/SBS/assets/530657/2436154b-2d54-4051-ba20-9d5dd67fbba7)

**Comment by thijskh** [comment]
Probably better to just not support the rexeps at all and ensure KNAW just lists all possible combinations? Regexp scopes are discouraged.

**Comment by baszoetekouw** [comment]
I think that if this occurs in eduGAIN, we need to support it, ugly or not :(

+1 for moving away from regexped scoped for SURFconext IdPs.

**Comment by mrvanes** [comment]
Is it ok that an _arbitrary_ knaw.nl displayname is chosen?

![Image](https://github.com/user-attachments/assets/81f8f6b4-28cf-492d-a76c-b8518f23f2a7)

**Comment by oharsta** [comment]
We have no more information to go-at, as we only have the regular expression to test against the domain name.

**Comment by baszoetekouw** [comment]
@logan-life @FlorisFokkinga what do you think?

**Comment by baszoetekouw** [comment]
The problems is that multiple IdPs assert the same `knaw.nl` schib_scope.  So how should we determine which one to use?

**Comment by baszoetekouw** [comment]
The real solution is probably to not rely on the shib_scope at all, and instead use the surf_crm_id to match the user's home institution.  Of course, that depends on moving away from EduTEAMS.

**Comment by mrvanes** [comment]
Of course, a human intervention in choosing the best match (as often as need when scopes change) would help, but probably is way too prohibitive and labour intensive?

**Comment by FlorisFokkinga** [comment]
The more specific domain should be selected over the regex, e.g., pick `nin.knaw.nl` over `.*\\.knaw\\.nl$`.

**Comment by oharsta** [comment]
@FlorisFokkinga We only fallback to the regular expressions (which is a discouraged feature) if there is no exact match for the domain / schac_home in the metadata feed. We can't treat regular domain names (e.g. non regular expressions) in the metadata feed as regular expressions, for this is error-prone. My advice is to close this issue.

**Comment by FlorisFokkinga** [comment]
> Is it ok that an _arbitrary_ knaw.nl displayname is chosen?
> 
> ![Image](https://private-user-images.githubusercontent.com/1901782/350370683-81f8f6b4-28cf-492d-a76c-b8518f23f2a7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM4MDIwNTYsIm5iZiI6MTcyMzgwMTc1NiwicGF0aCI6Ii8xOTAxNzgyLzM1MDM3MDY4My04MWY4ZjZiNC0yOGNmLTQ5MmQtYTc2Yy1iODUxOGYyM2YyYTcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgxNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MTZUMDk0OTE2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODI3MzJlYjA5YzNhYzM3ZTUxMDg0ZjU5NDJkYzM0M2QzOTM2NzA3ZjhlMWYyMmEzYTNhNDgxMGRjODA1YTU4NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.VtTlkUhktY038trDEbXvaDW9kH_d7axiIcVX5sDsEqw)

Testing with domain `nin.knaw.nl`, it works as expected: the name 'Netherlands Institute for Neuroscience (KNAW)' is shown.

But testing with `knaw.nl`, the name 'Hubrecht Institute & Westerdijk Fungal Biodiversity Institute (KNAW)' is shown, and not 'Koninklijke Nederlandse Akademie van Wetenschappen (KNAW)', as expected.

The problem seems that both KNAW proper, through the regex, and 'Hubrecht Institute (...)', directly, claim the `knaw.nl` domain.

It seems to me, SBS is behaving as expected, but the contents of the metadata is 'unexpected'.

**Comment by oharsta** [comment]
The metadata feed contains two occurrences of `<shibmd:Scope regexp="false">knaw.nl</shibmd:Scope>`. Both the "_Hubrecht Institute &amp; Westerdijk Fungal Biodiversity Institute (KNAW)_" and "Koninklijke Nederlandse Akademie van Wetenschappen (KNAW)" claim this domain. The parsing algorithm becomes indeterminant because of this.

**Comment by thijskh** [comment]
You should not be looking up the org name from the scopes as this is not what the attribute is meant for so can/will lead to incorrect results.

**Comment by baszoetekouw** [comment]
Looking at the code, the cause of the issue is that the scope is simply overwritten by the last value in the metadata file: https://github.com/SURFscz/SBS/blob/main/server/cron/idp_metadata_parser.py#L78-L79

Maybe the easiest solution is to add an override to the config, like this:
```yaml
metadata:
  idp_url: "file://data/idps-metadata.xml"
  parse_at_startup: False
  scope_override:
    knaw.nl: "http://adfs.knaw.nl/adfs/services/trust"
```
and then use this to override the scopes as parsed from the metadata file.

**Comment by oharsta** [comment]
In the branch https://github.com/SURFscz/SBS/tree/feature/refactor-login-process-883 we don't use the entity_id anymore to lookup the schac_home, as we dropped support for SSID. It is much easier to fix this, when this branch is merged into main. For now block.

**Comment by mrvanes** [comment]
Checked
